File: packages/permetic-web/test/contract-parity.test.ts
import { describe, expect, it } from 'vitest';
import { Project } from 'ts-morph';
import path from 'node:path';
import manifest from '../contract/manifest.json';

/**
 * TS-side half of the contract-freeze parity check. Both this test and
 * `ContractParityTest.kt` (JVM) assert against the same
 * `packages/permetic-web/contract/manifest.json` — change `index.d.ts` alone, or the
 * Kotlin mirror alone, without updating the manifest, and the side that didn't
 * change fails its own test. See spec 01, task 1.
 */

const project = new Project();
const sourceFile = project.addSourceFileAtPath(
  path.resolve(import.meta.dirname, '../src/index.d.ts'),
);

function getCapabilityNames(): string[] {
  const typeAlias = sourceFile.getTypeAliasOrThrow('CapabilityName');
  return typeAlias
    .getType()
    .getUnionTypes()
    .map((t) => t.getLiteralValueOrThrow() as string);
}

function getMethodNames(interfaceName: string): string[] {
  return sourceFile
    .getInterfaceOrThrow(interfaceName)
    .getMethods()
    .map((m) => m.getName());
}

const interfaceByCapability: Record<string, string> = {
  auth: 'AuthCapability',
  push: 'PushCapability',
  billing: 'BillingCapability',
  background: 'BackgroundCapability',
  storage: 'StorageCapability',
  system: 'SystemCapability',
};

describe('contract parity: index.d.ts vs manifest.json', () => {
  it('CapabilityName union matches the manifest keys', () => {
    expect(new Set(getCapabilityNames())).toEqual(new Set(Object.keys(manifest)));
  });

  it('every manifest key has a matching *Capability interface mapped', () => {
    expect(new Set(Object.keys(manifest))).toEqual(new Set(Object.keys(interfaceByCapability)));
  });

  for (const [capability, interfaceName] of Object.entries(interfaceByCapability)) {
    it(`${capability} methods match the manifest`, () => {
      const methods = getMethodNames(interfaceName);
      const expected = (manifest as Record<string, string[]>)[capability];
      expect(new Set(methods)).toEqual(new Set(expected));
    });
  }
});

File: packages/permetic/src/test/kotlin/ac/onyx/permetic/capability/ContractParityTest.kt
package ac.onyx.permetic.capability

import kotlinx.serialization.json.Json
import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * Kotlin-side half of the contract-freeze parity check. Both this test and
 * `packages/permetic-web/test/contract-parity.test.ts` assert against the same
 * `packages/permetic-web/contract/manifest.json` — change `index.d.ts` alone, or
 * [Contract]/[CapabilityName] alone, without updating the manifest, and the side
 * that didn't change fails its own test. See spec 01, task 1.
 */
class ContractParityTest {
    private val manifest: Map<String, List<String>> by lazy {
        val file = File("../permetic-web/contract/manifest.json")
        check(file.exists()) { "contract manifest not found at ${file.absolutePath}" }
        Json.decodeFromString(file.readText())
    }

    @Test
    fun `capability names match the manifest`() {
        assertEquals(manifest.keys, CapabilityName.entries.map { it.contractName }.toSet())
    }

    @Test
    fun `auth methods match the manifest`() = assertCapability(CapabilityName.AUTH, Contract.AUTH)

    @Test
    fun `push methods match the manifest`() = assertCapability(CapabilityName.PUSH, Contract.PUSH)

    @Test
    fun `billing methods match the manifest`() =
        assertCapability(
            CapabilityName.BILLING,
            Contract.BILLING,
        )

    @Test
    fun `background methods match the manifest`() =
        assertCapability(CapabilityName.BACKGROUND, Contract.BACKGROUND)

    @Test
    fun `storage methods match the manifest`() =
        assertCapability(
            CapabilityName.STORAGE,
            Contract.STORAGE,
        )

    @Test
    fun `system methods match the manifest`() =
        assertCapability(
            CapabilityName.SYSTEM,
            Contract.SYSTEM,
        )

    @Test
    fun `Contract byCapability covers every CapabilityName entry`() {
        assertEquals(CapabilityName.entries.toSet(), Contract.byCapability.keys)
    }

    private fun assertCapability(
        name: CapabilityName,
        methods: List<String>,
    ) {
        assertEquals(manifest.getValue(name.contractName).toSet(), methods.toSet())
    }
}
